csrc/nvshmem_binding.cu

```diff
@@ -0,0 +1,126 @@
+    if (e != cudaSuccess) {                                                                 \
+      printf("Failed: Cuda error %s:%d '%s'\n", __FILE__, __LINE__, cudaGetErrorString(e)); \
+      exit(EXIT_FAILURE);                                                                   \
+    }                                                                                       \
+  } while (0)
+
+template <typename T>
+T* mallocZeroBuffer(size_t size) {
+  T* ptr;
+  CUDACHECK(cudaMalloc(&ptr, size * sizeof(T)));
+  cudaMemset(ptr, 0, size * sizeof(T));
+  return ptr;
+}
+
+inline int get_sm_count() {
+  int device;
+  CUDACHECK(cudaGetDevice(&device));
+  int numSMs;
+  CUDACHECK(cudaDeviceGetAttribute(&numSMs, cudaDevAttrMultiProcessorCount, device));
+
+  return numSMs;
+}
+
+#define NVSHMEMCHECK(stmt)                                                                    \
+  do {                                                                                        \
+    int result = (stmt);                                                                      \
+    if (NVSHMEMX_SUCCESS != result) {                                                         \
+      fprintf(stderr, "[%s:%d] nvshmem failed with error %d \n", __FILE__, __LINE__, result); \
+      exit(-1);                                                                               \
+    }                                                                                         \
+  } while (0)
+
+namespace {
+
+at::Tensor get_unique_id() {
+  nvshmemx_uniqueid_t uid = NVSHMEMX_UNIQUEID_INITIALIZER;
+  nvshmemx_get_uniqueid(&uid);
+  return at::from_blob(&uid, sizeof(uid), at::kByte).clone();
+}
+
+int64_t unique_id_size() { return sizeof(nvshmemx_uniqueid_t); }
+
+int64_t init(at::Tensor uid, int64_t rank, int64_t world_size) {
+  TORCH_CHECK(uid.device().is_cpu(), "uid must be a CPU tensor");
+  TORCH_CHECK(uid.scalar_type() == at::kByte, "uid must be a byte tensor");
+  TORCH_CHECK(uid.numel() == sizeof(nvshmemx_uniqueid_t), "Invalid unique id size (expected ",
+              sizeof(nvshmemx_uniqueid_t), ", got ", uid.numel(), ")");
+  nvshmemx_uniqueid_t id;
+  std::memcpy(&id, uid.data_ptr(), sizeof(id));
+  nvshmemx_init_attr_t attr = NVSHMEMX_INIT_ATTR_INITIALIZER;
+  nvshmemx_set_attr_uniqueid_args(rank, world_size, &id, &attr);
+  return nvshmemx_init_attr(NVSHMEMX_INIT_WITH_UNIQUEID, &attr);
+}
+
+void finalize() { nvshmem_finalize(); }
+
+int64_t my_pe() { return nvshmem_my_pe(); }
+
+int64_t n_pes() { return nvshmem_n_pes(); }
+
+at::Tensor malloc_tensor(const std::vector<int64_t>& shape, c10::ScalarType dtype,
+                         const c10::Device& device) {
+  size_t size = c10::elementSize(dtype) * c10::multiply_integers(shape);
+  void* ptr = nvshmem_malloc(size);
+  if (ptr == nullptr) {
+    AT_ERROR("nvshmem_malloc failed. size: ", size);
+  }
+  return at::from_blob(
```

Comment:
We should add a comment here and on the python side to remind caller that `malloc_tensor` is a collective operation that needs to be called by all participants.

Also, caller should explicitly `del tensor` to guarantee the ordering of `nvshmem_free` instead of relying on GC. 

> The nvshmem_malloc, nvshmem_free, and nvshmem_align routines are collective operations that require participation by all PEs.
>
> https://docs.nvidia.com/nvshmem/api/gen/api/memory.html#nvshmem-malloc-nvshmem-free-nvshmem-align